action/protocol/staking/handler_stake_migrate.go

```diff
@@ -0,0 +1,208 @@
+		}
+	}
+	// clear candidate's self stake if the
+	if cand.SelfStakeBucketIdx == bucket.Index {
+		cand.SelfStake = big.NewInt(0)
+		cand.SelfStakeBucketIdx = candidateNoSelfStakeBucketIndex
+	}
+	if err := csm.Upsert(cand); err != nil {
+		return nil, nil, csmErrorToHandleError(cand.GetIdentifier().String(), err)
+	}
+	// update withdrawer balance
+	if err := withdrawer.AddBalance(bucket.StakedAmount); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to add balance %s", bucket.StakedAmount)
+	}
+	// put updated withdrawer's account state to trie
+	actionCtx := protocol.MustGetActionCtx(ctx)
+	if err := accountutil.StoreAccount(csm.SM(), actionCtx.Caller, withdrawer); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to store account %s", actionCtx.Caller.String())
+	}
+	// create receipt log
+	actLog := newReceiptLog(p.addr.String(), HandleWithdrawStake, protocol.MustGetFeatureCtx(ctx).NewStakingReceiptFormat)
+	actLog.AddTopics(byteutil.Uint64ToBytesBigEndian(bucket.Index), bucket.Candidate.Bytes())
+	actLog.AddAddress(actionCtx.Caller)
+	actLog.SetData(bucket.StakedAmount.Bytes())
+	return actLog, &action.TransactionLog{
+		Type:      iotextypes.TransactionLogType_WITHDRAW_BUCKET,
+		Amount:    bucket.StakedAmount,
+		Sender:    address.StakingBucketPoolAddr,
+		Recipient: actionCtx.Caller.String(),
+	}, nil
+}
+
+func (p *Protocol) ConstructExecution(ctx context.Context, act *action.MigrateStake, sr protocol.StateReader) (*action.Execution, error) {
+	sm := protocol.NewStateManagerWrapper(sr)
+	csm, err := p.constructCandidateStateManager(ctx, sm)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to construct state managers")
+	}
+	bucket, err := p.fetchBucket(csm, act.BucketIndex())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch bucket")
+	}
+	candidate := csm.GetByIdentifier(bucket.Candidate)
+	if candidate == nil {
+		return nil, errCandNotExist
+	}
+	duration := uint64(bucket.StakedDuration / p.helperCtx.BlockInterval(protocol.MustGetBlockCtx(ctx).BlockHeight))
+
+	return p.constructExecution(candidate.GetIdentifier(), bucket.StakedAmount, duration, act.Nonce(), act.GasLimit(), act.GasPrice())
+}
+
+func (p *Protocol) constructExecution(candidate address.Address, amount *big.Int, duration uint64, nonce uint64, gasLimit uint64, gasPrice *big.Int) (*action.Execution, error) {
+	contractAddress := p.config.MigrateContractAddress
+	data, err := StakingContractABI.Pack(
+		"stake0",
+		big.NewInt(int64(duration)),
+		common.BytesToAddress(candidate.Bytes()),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to pack data for contract call")
+	}
+	return action.NewExecution(
+		contractAddress,
+		nonce,
+		amount,
+		gasLimit,
+		gasPrice,
+		data,
+	)
+}
+
+func (p *Protocol) createNFTBucket(ctx context.Context, exeAct *action.Execution, sm protocol.StateManager) (*action.Receipt, error) {
+	exctPtl := execution.FindProtocol(protocol.MustGetRegistry(ctx))
```

Comment:
this part can move to `Start()` to get the `exctPtl` and keep it, and directly use it here